gcloud-java-storage/src/main/java/com/google/gcloud/storage/CopyWriter.java

```diff
@@ -0,0 +1,269 @@
+ */
+
+package com.google.gcloud.storage;
+
+import static com.google.gcloud.RetryHelper.runWithRetries;
+
+import com.google.common.base.MoreObjects;
+import com.google.gcloud.Restorable;
+import com.google.gcloud.RestorableState;
+import com.google.gcloud.RetryHelper;
+import com.google.gcloud.spi.StorageRpc;
+import com.google.gcloud.spi.StorageRpc.RewriteRequest;
+import com.google.gcloud.spi.StorageRpc.RewriteResponse;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+
+/**
+ * Google Storage blob copy writer. This class holds the result of a copy request.
+ * If source and destination blobs do not share the same location or storage class more than one
+ * RPC request is needed to copy the blob otherwise one or more {@link #copyChunk()} calls are
+ * necessary to complete the copy.
+ *
+ * @see <a href="https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite">Rewrite</a>
+ */
+public class CopyWriter implements Restorable<CopyWriter> {
+
+  private final StorageOptions serviceOptions;
+  private final StorageRpc storageRpc;
+  private RewriteResponse rewriteResponse;
+
+  CopyWriter(StorageOptions serviceOptions, RewriteResponse rewriteResponse) {
+    this.serviceOptions = serviceOptions;
+    this.rewriteResponse = rewriteResponse;
+    this.storageRpc = serviceOptions.rpc();
+  }
+
+  /**
+   * Returns the updated information for the just written blob. This method might block and issue
+   * several RPC requests to complete blob copy.
+   *
+   * @throws StorageException upon failure
+   */
+  public BlobInfo result() {
+    while(!isDone()) {
+      copyChunk();
+    }
+    return BlobInfo.fromPb(rewriteResponse.result);
+  }
+
+  /**
+   * Size of the blob being copied.
+   */
+  public long blobSize() {
+    return rewriteResponse.blobSize;
+  }
+
+  /**
+   * Returns {@code true} of blob rewrite finished, {@code false} otherwise.
+   */
+  public boolean isDone() {
+    return rewriteResponse.isDone;
+  }
+
+  /**
+   * Returns the number of bytes copied. 
+   */
+  public long totalBytesCopied() {
+    return rewriteResponse.totalBytesRewritten;
+  }
+
+  /**
+   * Copies the next chunk of the blob. An RPC is issued only if copy has not finished yet
+   * ({@link #isDone} returns {@code false}).
+   *
+   * @throws StorageException upon failure
+   */
+  public void copyChunk() {
+    if (!isDone()) {
+      try {
+        this.rewriteResponse = runWithRetries(new Callable<RewriteResponse>() {
+          @Override
+          public RewriteResponse call() {
+            return storageRpc.continueRewrite(rewriteResponse);
+          }
+        }, serviceOptions.retryParams(), StorageImpl.EXCEPTION_HANDLER);
+      } catch (RetryHelper.RetryHelperException e) {
+        throw StorageException.translateAndThrow(e);
+      }
+    }
+  }
+
+  @Override
+  public RestorableState<CopyWriter> capture() {
+    return StateImpl.builder(
```

Comment:
Fine as is, but I am not sure the Builder in this case giving us much as oppose to just
passing `this` to `StateImpl` constructor.